### PR TITLE
Optimize thanos sidecar code logic and add `TracingConfigFile` field.

### LINKED ISSUE #3763: Optimize thanos sidecar code logic and add `TracingConfigFile` field
**Why do we need it?**
1. Optimize`ObjectStorageConfig` and `ObjectStorageConfigFile` field processing logic , which match comment logic `When used alongside with ObjectStorageConfig, ObjectStorageConfigFile takes precedence`.
2. Add `TracingConfigFile` field support `-tracing.config-file` args.
3. When disable prometheus compact,we can set the `--storage.tsdb.min-block-duration` and `--storage.tsdb.max-block-duration `parameters together.

**Environment**
master branch.

### PR DESCRIPTION
**Pull request by mikechengwei**
Fix #3763
1. Optimize`ObjectStorageConfig` and `ObjectStorageConfigFile` field processing logic , which match comment logic `When used alongside with ObjectStorageConfig, ObjectStorageConfigFile takes precedence`.
2. Add `TracingConfigFile` field support `-tracing.config-file` args.
3. When disable prometheus compact,we can set the `--storage.tsdb.min-block-duration` and `--storage.tsdb.max-block-duration `parameters together.

- release-note:enhancement
```release-note:REPLACEME
Add `TracingConfigFile` field into thanos configuration.
```

### CODE REVIEW

File: pkg/prometheus/statefulset.go [hunk 1]
```diff
@@ -750,7 +750,9 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
 		}
 
 		if p.Spec.Thanos.ObjectStorageConfig != nil || p.Spec.Thanos.ObjectStorageConfigFile != nil {
-			if p.Spec.Thanos.ObjectStorageConfig != nil {
+			if p.Spec.Thanos.ObjectStorageConfigFile != nil {
+				container.Args = append(container.Args, "--objstore.config-file="+*p.Spec.Thanos.ObjectStorageConfigFile)
+			} else {
```
Comment (s-urbaniak) [inline pkg/prometheus/statefulset.go:755]: this feels like a duplication of logic that is already present in Thanos itself: https://github.com/thanos-io/thanos/blob/dd670a3e8cd2d99bd86169bdaf6bf304ab5ea2f9/pkg/extflag/pathorcontent.go#L50-L73

As far as i understand we don't really need the if-else switch.

File: pkg/prometheus/statefulset.go [hunk 2]
```diff
@@ -779,14 +776,18 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
...
-			container.Env = append(container.Env, v1.EnvVar{
-				Name: "TRACING_CONFIG",
-				ValueFrom: &v1.EnvVarSource{
-					SecretKeyRef: p.Spec.Thanos.TracingConfig,
-				},
-			})
+		if p.Spec.Thanos.TracingConfig != nil || p.Spec.Thanos.TracingConfigFile != nil {
+			if p.Spec.Thanos.TracingConfigFile != nil {
+				container.Args = append(container.Args, "--tracing.config-file="+*p.Spec.Thanos.TracingConfigFile)
+			} else {
```
Comment (s-urbaniak) [inline pkg/prometheus/statefulset.go:782]: see above regarding the if-else construct.

File: pkg/prometheus/statefulset.go [hunk 3]
```diff
@@ -807,6 +808,7 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
 	}
 	if disableCompaction {
 		promArgs = append(promArgs, "--storage.tsdb.max-block-duration=2h")
+		promArgs = append(promArgs, "--storage.tsdb.min-block-duration=2h")
```
Comment (s-urbaniak) [inline pkg/prometheus/statefulset.go:811]: this looks correct to me but leaving @brancz and/or @simonpasquier to have another look.

**Comment by s-urbaniak** [comment]
thank you @mikechengwei for the contribution! this generally looks good to me, i just have a nit regardin the if-else construct and a follow-up question for the prometheus settings.

File: pkg/prometheus/statefulset.go (hunk 1 above)
Comment (mikechengwei) [reply pkg/prometheus/statefulset.go:755]: Ok.

File: pkg/prometheus/statefulset.go (hunk 2 above)
Comment (mikechengwei) [reply pkg/prometheus/statefulset.go:782]: Ok.

File: pkg/prometheus/statefulset.go (hunk 1 above)
Comment (simonpasquier) [reply pkg/prometheus/statefulset.go:755]: From what I can tell, Thanos won't start if both flags are defined and not empty:
https://github.com/thanos-io/thanos/blob/dd670a3e8cd2d99bd86169bdaf6bf304ab5ea2f9/pkg/extflag/pathorcontent.go#L53-L55

But the [help text](https://github.com/thanos-io/thanos/blob/dd670a3e8cd2d99bd86169bdaf6bf304ab5ea2f9/pkg/extflag/pathorcontent.go#L36) for the `*-file` flag is a bit misleading: it says `Alternative to '...' flag (lower priority). ...`. It should either say that both flags are exclusive or implement the priority logic.

File: pkg/prometheus/statefulset.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/prometheus/statefulset.go:811]: The default value for `--storage.tsdb.min-block-duration` is already [2h](https://github.com/prometheus/prometheus/blob/063154eab720d8c3d495bd78312c0df090d0bf23/cmd/prometheus/main.go#L197-L198) and it shouldn't change for 2.x versions but it doesn't hurt to set it explicitly.

**Review by simonpasquier — COMMENTED** [review]
Process-wise, I'd prefer to have a different PR for each change (e.g. object storage config, tracing config and block duration).

File: pkg/prometheus/statefulset.go (hunk 1 above)
Comment (mikechengwei) [reply pkg/prometheus/statefulset.go:755]: I think we can add if-else to keep Semantic correctness.

**Comment by mikechengwei** [comment]
> Process-wise, I'd prefer to have a different PR for each change (e.g. object storage config, tracing config and block duration).

Do I need to split the PR and submit it in turn this time?

**Review by blaquezzz — APPROVED** [review]

File: pkg/apis/monitoring/v1/types.go [hunk 4]
```diff
@@ -537,6 +537,9 @@ type ThanosSpec struct {
 	ListenLocal bool `json:"listenLocal,omitempty"`
 	// TracingConfig configures tracing in Thanos. This is an experimental feature, it may change in any upcoming release in a breaking way.
 	TracingConfig *v1.SecretKeySelector `json:"tracingConfig,omitempty"`
+	// TracingConfig specifies the path of the tracing configuration file.
+	// When used alongside with TracingConfig, TracingConfigFile takes precedence.
+	TracingConfigFile *string `json:"tracingConfigFile,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:542]: I'd rather have a `string` here than a `*string`.

**Review by simonpasquier — COMMENTED** [review]
I have one small request but fine if it got merged as-is. Leaving final call to @lilic as the next release shepherd :)

**Comment by lilic** [comment]
Thanks for the ping Simon, I plan on doing the release tomorrow morning. But agreed on Simons comment lets address that first.

File: pkg/apis/monitoring/v1/types.go (hunk 4 above)
Comment (mikechengwei) [reply pkg/apis/monitoring/v1/types.go]: done

**Comment by mikechengwei** [comment]
@simonpasquier  I have modified，PTAL.

**Review by simonpasquier — APPROVED** [review]
lgtm

**Comment by mikechengwei** [comment]
@lilic Please help me merge this pr.

**Comment by mikechengwei** [comment]
@simonpasquier Can you help me merge this pr?

**Review by lilic — APPROVED** [review]
Thanks, lgtm

**Pull request merged**